Refactor Path so that it works on windows

a20012251 opened the pull request:
Even though this diff is large, it's rather trivial.

The Path.swift file was assuming POSIX, and because we want to support Windows, we need to remove this assumption.

I'm adding support for standard windows paths of the form DRIVE_LETTER:\abc\...\xyz. More complex NT path are not worth support for the time being. The code only has 3 `#if os`, which means that most of it is quite portable.

I had to remove the Path.root constant, because it doesn't make sense in a cross-platform environment. People will have to be explicit about "/" when they assume a posix context.

All tests pass

There are some mechanisms in SwiftPM that deal with in-memory filesystems. These flows will break on Windows, so we'll have to fix them eventually, probably by creating a virtual filesystem and forcing posix on it via its constructor.

jmittert commented:
I believe you've missed a `.root` in `Sources/Workspace/Destination.swift`

aciidb0mb3r commented on Sources/Basic/Path.swift:
```diff
@@ -28,14 +29,23 @@
+#if os(Windows)
+public let PATH_SEPARATOR = "\\"
+#else
+public let PATH_SEPARATOR = "/"
```
I think this should be a static property on AbolsutePath so it's correctly namespaced: `AbsolutePath.pathSeparator`

aciidb0mb3r commented on Sources/Basic/Path.swift:
```diff
@@ -531,32 +542,60 @@ extension AbsolutePath {
     return false
 }
 
+extension Character {
```
Please make these extensions `private`.

aciidb0mb3r commented:
I think this is great! I left some minor comments inline. We should also add some unit tests for Windows especially where we take a different code path.

a20012251 commented:
Addressed comments.
I stopped working on this because I need to do another project, so maybe Gwen could write tests or I can write them but in some weeks :(

gmittert commented on Sources/Basic/Path.swift:
```diff
@@ -531,32 +542,60 @@ extension AbsolutePath {
 
+extension Character {
+    var isAlpha: Bool {
+        return (self >= "a" && self <= "Z") || (self >= "A" && self <= "Z")
```
I believe this should be lowercase "z" for the first comparison.

tomerd commented:
@compnerd is this still relevant given your recent work for Windows support?

compnerd commented:
There are some interesting ideas here.  There is something to be gained from this, but I need to look more carefully at the changes (at the very least, the use explicit references to root actually would really help the current tree).

shahmishal commented:
We moved our default branch to `main` and deleted `master` branch, GitHub automatically closed the PR. Please update the base branch to `main`.  More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412

tomerd commented:
Hi @a20012251 and @compnerd, the Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please rebase on top of the `main` and we can reopen.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412
